crates/chimney-core/src/tls/cache.rs

```diff
@@ -0,0 +1,266 @@
+    // Try to make it relative to current directory, otherwise just show the file name
+    std::env::current_dir()
+        .ok()
+        .and_then(|cwd| full_path.strip_prefix(&cwd).ok())
+        .map(|p| p.display().to_string())
+        .unwrap_or_else(|| {
+            full_path
+                .file_name()
+                .map(|n| n.to_string_lossy().to_string())
+                .unwrap_or_else(|| full_path.display().to_string())
+        })
+}
+
+/// Create the certificate directory for a site
+pub fn create_cert_directory(site_name: &str, cert_dir: &Path) -> Result<PathBuf, ServerError> {
+    // Validate site name to prevent path traversal
+    validate_site_name(site_name)?;
+
+    let site_cert_dir = cert_dir.join(site_name);
+
+    fs::create_dir_all(&site_cert_dir).map_err(|e| {
+        ServerError::CertificateDirectoryCreationFailed {
+            path: safe_display_path(&site_cert_dir),
+            message: e.to_string(),
+        }
+    })?;
+
+    return Ok(site_cert_dir);
+}
+
+/// Save certificate and key to disk (atomic write)
+pub fn save_certificate(
+    site_name: &str,
+    cert_dir: &Path,
+    cert_pem: &[u8],
+    key_pem: &[u8],
+) -> Result<(), ServerError> {
+    // create_cert_directory already validates site_name
+    let site_cert_dir = create_cert_directory(site_name, cert_dir)?;
+
+    let cert_path = site_cert_dir.join("cert.pem");
+    let key_path = site_cert_dir.join("key.pem");
+
+    // Write certificate
+    let temp_cert = site_cert_dir.join(".cert.pem.tmp");
+    fs::write(&temp_cert, cert_pem).map_err(|e| {
+        ServerError::TlsInitializationFailed(format!("Failed to write certificate: {e}"))
+    })?;
+    fs::rename(&temp_cert, &cert_path).map_err(|e| {
+        ServerError::TlsInitializationFailed(format!("Failed to move certificate: {e}"))
+    })?;
+
+    // Write private key with restricted permissions
+    let temp_key = site_cert_dir.join(".key.pem.tmp");
+    fs::write(&temp_key, key_pem).map_err(|e| {
+        ServerError::TlsInitializationFailed(format!("Failed to write private key: {e}"))
+    })?;
+
+    // Set restrictive permissions on private key
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        let mut perms = fs::metadata(&temp_key)
+            .map_err(|e| {
+                ServerError::TlsInitializationFailed(format!("Failed to get key file permissions: {e}"))
+            })?
+            .permissions();
+        perms.set_mode(0o600);  // Owner read/write only
+        fs::set_permissions(&temp_key, perms).map_err(|e| {
+            ServerError::TlsInitializationFailed(format!("Failed to set key file permissions: {e}"))
+        })?;
+    }
+
+    // NOTE: On Windows, file permissions are managed via ACLs and require
+    // additional dependencies (winapi). Consider using proper ACLs for production.
+    // For now, Windows users should ensure proper NTFS permissions manually.
+
+    fs::rename(&temp_key, &key_path).map_err(|e| {
```

Comment:
The atomic write pattern is good, but if `fs::rename()` fails, the temporary file (`.key.pem.tmp`) will be left on disk. Consider adding cleanup:

```rust
fs::rename(&temp_key, &key_path).map_err(|e| {
    // Clean up temp file on failure
    let _ = fs::remove_file(&temp_key);
    ServerError::TlsInitializationFailed(format!("Failed to move private key: {e}"))
})?;
```

Also applies to the certificate temp file above.